lib/jormungandr/src/Cardano/Wallet/Binary/Jormungandr.hs

```diff
@@ -8,5 +11,178 @@
 -- The format is for the Shelley era as implemented by the Jörmungandr node.
 
 module Cardano.Wallet.Binary.Jormungandr
-    (
+    ( getBlockHeader
+    , Message (..)
+    , BlockHeader (..)
+
+     -- * Re-export
+    , runGet
+
+    -- * For dev
+    , genesisBlock
+    , dev
     ) where
+
+import Prelude
+
+import Cardano.Wallet.Primitive.Types
+    ( SlotId (..) )
+import Data.Binary.Get
+    ( Get
+    , getByteString
+    , getWord16be
+    , getWord32be
+    , getWord8
+    , isEmpty
+    , isolate
+    , runGet
+    , skip
+    )
+import Data.ByteArray.Encoding
+    ( Base (Base16), convertFromBase )
+import Data.ByteString
+    ( ByteString )
+import Data.Word
+    ( Word16, Word32 )
+
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy as BL
+
+
+-- | Messages is what the block body consists of.
+--
+-- Every message is prefixed with a message header.
+--
+--  Following, as closely as possible:
+-- https://github.com/input-output-hk/rust-cardano/blob/e0616f13bebd6b908320bddb1c1502dea0d3305a/chain-impl-mockchain/src/message/mod.rs#L22-L29
+data Message
+    = Initial [ConfigParam]
+    | OldUtxoDeclaration TODO
+    | Transaction TODO
+    | Certificate TODO
+    | UpdateProposal SignedUpdateProposal
+    | UpdateVote SignedVote
+    | UnimplementedMessage Int -- For development. Remove later.
+    deriving Show
+
+
+data BlockHeader = BlockHeader
+    { version :: Word16
+    , contentSize :: Word32
+    , slot :: SlotId
+    , chainLength :: Word32
+    , contentHash :: ByteString
+    , parentHeaderHash :: Maybe ByteString
+    } deriving (Show, Eq)
+
+-- |
+--
+-- Header
+--
+data Block = Block BlockHeader [Message]
+    deriving Show
+
+
+data SignedUpdateProposal = SignedUpdateProposal
+    deriving Show
+data TODO = TODO
+    deriving Show
+data SignedVote = SignedVote
+    deriving Show
+data ConfigParam = ConfigParam
+    deriving Show
+
+{-# ANN getBlockHeader ("HLint: ignore Use <$>" :: String) #-}
```

Comment:
Didn't want to use applicative syntax. Wanted to keep the binary format separate from the `BlockHeader` constructor and the ordering of its arguments.